### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by mobile32**

### CODE REVIEW

File: app/controllers/users/confirmations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
Comment (wojtodzio) [inline app/controllers/users/confirmations_controller.rb]: ✂️The whole file is commented out - we don't need it, you can safely remove it 😄

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
Comment (wojtodzio) [inline app/controllers/users/omniauth_callbacks_controller.rb:3]: ✂️Not needed as well

File: app/controllers/users/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class Users::PasswordsController < Devise::PasswordsController
```
Comment (wojtodzio) [inline app/controllers/users/passwords_controller.rb]: ✂️Not needed as well

File: app/controllers/users/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+class Users::RegistrationsController < Devise::RegistrationsController
+  before_action :configure_sign_up_params, only: [:create]
+  before_action :configure_account_update_params, only: [:update]
+
+  # GET /resource/sign_up
```
Comment (wojtodzio) [inline app/controllers/users/registrations_controller.rb]: ✂️You can remove those comments lines, they were here only to suggest you what/how you could possibly change something in Devise, but now that we've successfully changed permitted parameters, they don't bring much value anymore

File: app/controllers/users/registrations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  # protected
+
+  # If you have extra params to permit, append them to the sanitizer.
+  def configure_sign_up_params
+    devise_parameter_sanitizer.permit(:sign_up, keys: %i[first_name last_name])
+  end
+
+  # If you have extra params to permit, append them to the sanitizer.
+  def configure_account_update_params
+    devise_parameter_sanitizer.permit(:account_update, keys: [:attribute])
```
Comment (wojtodzio) [inline app/controllers/users/registrations_controller.rb]: ✂️ We don't have update implemented yet, you can remove this callback

File: app/controllers/users/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Users::SessionsController < Devise::SessionsController
```
Comment (wojtodzio) [inline app/controllers/users/sessions_controller.rb]: ✂️Not needed as well

File: app/controllers/users/unlocks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::UnlocksController < Devise::UnlocksController
```
Comment (wojtodzio) [inline app/controllers/users/unlocks_controller.rb]: ✂️Not needed as well

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
```
Comment (wojtodzio) [inline app/models/user.rb]: ✂️You can remove those comments

File: app/views/devise/confirmations/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (wojtodzio) [inline app/views/devise/confirmations/new.html.erb]: ✂️We don't have emails confirmation - you can safely remove that file :)

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
```
Comment (wojtodzio) [inline app/views/devise/mailer/confirmation_instructions.html.erb]: ✂️We don't have emails confirmation - you can safely remove that file :)

File: app/views/devise/mailer/unlock_instructions.html.erb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @resource.email %>!</p>
```
Comment (wojtodzio) [inline app/views/devise/mailer/unlock_instructions.html.erb]: ✂️We don't have accounts auto locking - you can safely remove that file :)

File: app/views/devise/registrations/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
+
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
```
Comment (wojtodzio) [inline app/views/devise/registrations/edit.html.erb:3]: ✂️We don't have accounts edit right now - you can safely remove that file :)

File: app/views/devise/unlocks/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
```
Comment (wojtodzio) [inline app/views/devise/unlocks/new.html.erb]: ✂️We don't have accounts auto locking - you can safely remove that file :)

File: config/database.yml [hunk 14]
```diff
@@ -17,9 +17,12 @@
 default: &default
```
Comment (wojtodzio) [inline config/database.yml:1]: We prefer to keep only `config/database.yml.sample` (without password) under version control, and store `config/database.yml` only locally. We did it in the previous project :)

File: spec/features/auth_spec.rb [hunk 15]
```diff
@@ -0,0 +1,136 @@
+RSpec.feature 'User authentication' do
+  context 'With valid details' do
+    scenario 'sign up' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: `sign up` is a feature - and we prefer to describe a happy path inside a scenario description string. E.g. here you could describe it like that:
`scenario 'user signs up'`

File: spec/features/auth_spec.rb [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+      fill_in 'Password confirmation', with: 'jkpassword'
+
+      expect do
+        click_button 'Create account'
+      end.to change { User.count }.by(1)
+
+      expect(page).to have_content 'Your account has been successfully created'
+    end
+
+    scenario 'log into' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Describe happy path here as well :)

File: spec/features/auth_spec.rb [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+      sign_in user
+      visit '/'
+      click_link_or_button 'Log out'
+
+      expect(page).to have_content 'Log out successfully successfully'
+    end
+  end
+
+  context 'With invalid details' do
+    scenario 'blank fields while sign up' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Describe a path here as well :)
Something like: 'User cannot sign up without filling all required fields'

File: spec/features/auth_spec.rb [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+      expect(page).to have_error_messages "First name can't be blank",
+                                          "Last name can't be blank",
+                                          "Email can't be blank",
+                                          "Password can't be blank"
+
+      expect(current_path).to eq(user_registration_path)
+    end
+
+    scenario 'not same passwords' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Description

File: spec/features/auth_spec.rb [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+      expect do
+        click_button 'Create account'
+      end.to change { User.count }.by(0)
+
+      expect(page).to have_error_messages "Password confirmation does not match"
+      expect(current_path).to eq(user_registration_path)
+    end
+
+    scenario 'already registered email' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Description

File: spec/features/auth_spec.rb [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+      expect do
+        click_button 'Create account'
+      end.to change { User.count }.by(0)
+
+      expect(page).to have_error_messages 'Email has already been taken'
+      expect(current_path).to eq(user_registration_path)
+    end
+
+    scenario 'invalid email' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Description

File: spec/features/auth_spec.rb [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+      fill_in 'Password confirmation', with: 'jkpassword'
+
+      expect do
+        click_button 'Create account'
+      end.to change { User.count }.by(0)
+
+      expect(current_path).to eq(user_registration_path)
+    end
+
+    scenario 'too short password' do
```
Comment (wojtodzio) [inline spec/features/auth_spec.rb]: Description

**Review by wojtodzio — CHANGES_REQUESTED** [review]
A few small changes 😄

File: app/controllers/users/omniauth_callbacks_controller.rb (hunk 2 above)
Comment (mobile32) [reply app/controllers/users/omniauth_callbacks_controller.rb:3]: We need in next step for google auth

File: app/views/devise/registrations/edit.html.erb (hunk 12 above)
Comment (mobile32) [reply app/views/devise/registrations/edit.html.erb:3]: I use this view in next step

File: config/database.yml (hunk 14 above)
Comment (mobile32) [reply config/database.yml:1]: done

Comment (wojtodzio) [inline config/database.yml:1]: Now that we have `database.yml.sample` this file can be removed from version control

File: config/database.yml.sample [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+#
+default: &default
+  adapter: postgresql
+  encoding: unicode
+  host: localhost
+  # For details on connection pooling, see Rails configuration guide
+  # http://guides.rubyonrails.org/configuring.html#database-pooling
+  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+  username: postgres
+  password: haslo
```
Comment (wojtodzio) [inline config/database.yml.sample]: Please, do not include any passwords in the sample files (or any other files which are under version control)

File: config/database.yml.sample [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+
+development:
+  <<: *default
+  database: epic_gallery_development
+
+  # The specified database role being used to connect to postgres.
```
Comment (wojtodzio) [inline config/database.yml.sample]: ✂️ You can remove those comments

**Review by wojtodzio — CHANGES_REQUESTED** [review]
Few nitpicks left. You can merge it by yourself after fixing them :)

File: config/database.yml (hunk 14 above)
Comment (mobile32) [reply config/database.yml:1]: The file has been added to gitignore and removed.

File: config/database.yml.sample (hunk 22 above)
Comment (mobile32) [reply config/database.yml.sample]: done

File: config/database.yml.sample (hunk 23 above)
Comment (mobile32) [reply config/database.yml.sample]: done

**Pull request merged**